pkg/k8scfgmap/watch.go

```diff
@@ -0,0 +1,87 @@
+package k8scfgmap
+
+import (
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	k8sTest "k8s.io/client-go/testing"
+)
+
+var log = logging.LoggerForModule()
+
+// ConfigMapWatcher watches a config map in a given namespaces and evokes a callback function
+// when changes are detected.
+type ConfigMapWatcher struct {
+	k8sClient    kubernetes.Interface
+	modifiedFunc func(*v1.ConfigMap)
+}
+
+// NewConfigMapWatcher creates a new config map watcher.
+func NewConfigMapWatcher(k8sClient kubernetes.Interface, modifiedFunc func(*v1.ConfigMap)) *ConfigMapWatcher {
+	return &ConfigMapWatcher{k8sClient: k8sClient, modifiedFunc: modifiedFunc}
+}
+
+// Watch a config map in the given namespace bound by the context.
+func (w *ConfigMapWatcher) Watch(ctx concurrency.Waitable, namespace string, name string) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			watcher, err := w.k8sClient.CoreV1().ConfigMaps(namespace).Watch(
+				concurrency.AsContext(ctx),
+				metav1.SingleObject(metav1.ObjectMeta{Name: name, Namespace: namespace}))
+			if err != nil {
+				log.Errorw("Unable to create config map watcher", zap.Error(err))
+				continue
+			}
+			w.onChange(ctx, watcher.ResultChan())
+			watcher.Stop()
+		}
+	}
+}
+
+func (w *ConfigMapWatcher) onChange(ctx concurrency.Waitable, eventChannel <-chan watch.Event) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event, open := <-eventChannel:
+			if open {
+				switch event.Type {
```

Comment:
In the case the config map gets removed, what would actually happen with our metrics server 🤔

Reply:
If the client CA is still valid nothing, if it is invalid and cannot be updated then requests should be rejected.